### DISCUSSION TITLE
Creating a multi-file LibIGL project

### DISCUSSION THREAD

**Discussion by drathin4** [libigl/libigl#1804] [Q&A] [answered]
I'm have created my own LibIGL project and it had only 1 source file in it so far - `main.cpp`. But now, I want to make some changes to `mesh_boolean.cpp`. So I added that file too to the project folder. And I did `cmake ..` again in the project\build folder. And then I built it in Visual Studio. However it gave compilation issues. So I added all the dependent `.h` files to my project. Visual Studio is still giving compilation issues and asking me to add all the corresponding `.cpp` files.

So my question is for creating a multi-file project on top of LibIGL, do we have to add all the dependent `.h & .cpp` files to build the code? Or there is some way around where I can have only the needed files (in my case `main.cpp, mesh_boolean.cpp`) in the project and still build it correctly ?

**Comment by jdumas** [comment, marked as answer]
You need to use CMake to define your own project, e.g. take a look at our [example project template](https://github.com/libigl/libigl-example-project).

**Reply by drathin4** [reply]
I have already built my project based on the example project template. But the example project contains only 1 `main.cpp` file. Even in my case, with only 1 `main.cpp` file, it was building correctly. But the moment I added `mesh_boolean.cpp` to the project files (inside project directory), it had dependency issues. I had done `cmake ..` so that the solution is regenerated and it contains `mesh_boolean.cpp` as additional source file. But build had dependency issues.

**Reply by jdumas** [reply]
If it's a dependency issue, you need to enable the required LIBIGL_XXX option and add the `igl:::xxx` target to your project dependency in your `CMakeLists.txt`.

**Reply by drathin4** [reply]
I don't think the problem lies with CMakeLists.txt. I already have all that you mentioned in that.

cmake_minimum_required(VERSION 3.1)
project(example)

set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_CURRENT_SOURCE_DIR}/cmake)

# libigl
option(LIBIGL_WITH_OPENGL            "Use OpenGL"         ON)
option(LIBIGL_WITH_OPENGL_GLFW       "Use GLFW"           ON)
option(LIBIGL_WITH_CGAL              "Use CGAL"           ON)

find_package(LIBIGL REQUIRED QUIET)

# Add your project files
file(GLOB SRCFILES *.cpp)
add_executable(${PROJECT_NAME} ${SRCFILES})
target_link_libraries(${PROJECT_NAME} igl::core igl::opengl igl::opengl_glfw igl::cgal)
igl_copy_cgal_dll(${PROJECT_NAME})

When I meant build issue, I meant at VisualStudio level. For example, the build complains, `#include "mesh_boolean.h"` not found. That is the first code line of `mesh_boolean.cpp`. Subsequently, even if I add all the dependent .h files, there are still build issues in VisualStudio.

**Reply by jdumas** [reply]
You have to give more details about the exact error that you are facing. What is the error message you get from VS? If you are copying files from the tutorial, make sure you don't have two `main()` functions in your project or that will not work.

**Reply by drathin4** [reply]
The exact VS compilation error is:

Build started...
1>------ Skipped Build: Project: Continuous, Configuration: Debug x64 ------
1>Project not selected to build for this solution configuration 
2>------ Skipped Build: Project: Experimental, Configuration: Debug x64 ------
2>Project not selected to build for this solution configuration 
3>------ Skipped Build: Project: Nightly, Configuration: Debug x64 ------
3>Project not selected to build for this solution configuration 
4>------ Skipped Build: Project: NightlyMemoryCheck, Configuration: Debug x64 ------
4>Project not selected to build for this solution configuration 
5>------ Skipped Build: Project: RUN_TESTS, Configuration: Debug x64 ------
5>Project not selected to build for this solution configuration 
6>------ Build started: Project: example, Configuration: Debug x64 ------
6>mesh_boolean.cpp
6>C:\Rathin\Projects\MeshBoolean\mesh_boolean.cpp(10,10): fatal error C1083: Cannot open include file: 'mesh_boolean.h': No such file or directory
6>Done building project "example.vcxproj" -- FAILED.
7>------ Skipped Build: Project: ALL_BUILD, Configuration: Debug x64 ------
7>Project not selected to build for this solution configuration 
========== Build: 0 succeeded, 1 failed, 3 up-to-date, 6 skipped ==========

Subsequently, even if I add all the dependent .h files, there are still compilation issues in VS.

**Reply by jdumas** [reply]
Yes, that's because `"mesh_boolean.h"` is a relative path to the header. When you move this guy to your project, you have to use the full path, `<igl/copyleft/cgal/mesh_boolean.h>`

**Reply by drathin4** [reply]
I included the full path for all headers as you mentioned, but I'm still getting different compilation errors.

C:\Rathin\Projects\MeshBoolean\mesh_boolean.cpp(43,38): error C2995: 'bool igl::copyleft::cgal::mesh_boolean(const Eigen::MatrixBase<Derived> &,const Eigen::MatrixBase<U> &,const Eigen::MatrixBase<DerivedVB> &,const Eigen::MatrixBase<DerivedFB> &,const igl::MeshBooleanType &,Eigen::PlainObjectBase<DerivedVC> &,Eigen::PlainObjectBase<DerivedFC> &,Eigen::PlainObjectBase<DerivedJ> &)': function template has already been defined
C:\Rathin\Projects\MeshBoolean\libigl\include\igl\copyleft\cgal\mesh_boolean.h(51): message : see declaration of 'igl::copyleft::cgal::mesh_boolean'

I even tried including mesh_boolean.h header file in the project (and altering header paths), but the compilation error persists. How to resolve this error ? Is it something related to function template redefinition ?

**Reply by jdumas** [reply]
Yeah mean you cannot define the same function twice in a .cpp lib. The error message is pretty clear. Why are you copying code from `cgal/mesh_boolean.h` into your project anyway? Just include the header and use the function...

**Reply by drathin4** [reply]
I'm not copying the existing code in my project. I just want to include the `mesh_boolean.cpp` file in my project. I want to uncomment `//#define MESH_BOOLEAN_TIMING` since I need to check the performance. So, as you mentioned, I included that file and used full path for headers instead of relative path.

**Reply by jdumas** [reply]
You do not need to "include mesh_boolean.cpp" into your project. If you are linking against `igl::cgal` in your CMake file, then it is already a part of your project. If you want to add #define MESH_BOOLEAN_TIMING, please do so in your CMake project.

**Reply by drathin4** [reply]
I'm sorry, but I'm completely new to CMake as well. How to add preprocessor directives to CMake project ? Do we add them to the `CMakeLists.txt` file or while using the `cmake ..` command ? Could you please tell me the syntax as well ?

**Reply by jdumas** [reply]
Here are some resources to learn CMake:

- [How to CMake Good](https://www.youtube.com/playlist?list=PLK6MXr8gasrGmIiSuVQXpfFuE1uPT615s), video tutorials for beginners (Win/Linux) **[Beginner]**
- [How to create a Xcode project using CMake](https://www.youtube.com/watch?v=9EADCyN_Fek), video tutorial for beginners **[Beginner]**
- [Daniel Pfeifer “Effective CMake”](https://www.youtube.com/watch?v=bsXLMQ6WgIk), the first 45min are great! **[Intermediate/Advanced]**
- [An Introduction to Modern CMake](https://cliutils.gitlab.io/modern-cmake/), very well-organized booklet, see also section “Other sources” **[Intermediate]**
- [Effective Modern CMake](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1), great list of dos and don’ts **[Intermediate]**

**Reply by drathin4** [reply]
A couple of days ago, I had figured out how to add preprocessor directives to `CMakeLists.txt` file:
`add_compile_definitions(MESH_BOOLEAN_TIMING)`

But my question still remains about adding `mesh_boolean.cpp` file in my project or for that matter any other list of files. If I want to make some local code changes to files, I would need to include them in the project, right ? And then it will still give the function template redefinition error during build process. Then how to resolve such errors ? 

I'm assuming that we can make local code changes since LibIGL is open source.

**Reply by jdumas** [reply]
Look, you cannot have more than one definition of the same function linked in the same executable. This is called the [one-definition rule](https://en.wikipedia.org/wiki/One_Definition_Rule) and it's really at the heart of C++ programming.

If you want to use libigl in your project, or modify an existing function you should:
1. If you are copying the function into your own project, you need to use a different namespace/give the function a different name, to prevent collision with libigl.
2. If you are making a change to libigl, and you intend to contribute it back into libigl, then yes, edit directly the file in the libigl repository, but do not "copy" it into your project.

**Reply by drathin4** [reply]
OK, thanks. You can close this issue.
